### ISSUE TITLE
Invalid syntax error message

### ISSUE THREAD

**Issue by HolQue** [test-fullautomation/python-jsonpreprocessor#120]
Previously the test case JPP_0202-(PARAMETER_SUBSTITUTION)-[GOODCASE] caused the following error message:

`'Could not set variable 'list_variant_numbersVAR-${variant_str([0]}_MS-${milestone_'2')}_param' with value 'value'!'`

Based on:

https://github.com/test-fullautomation/python-jsonpreprocessor/blob/develop/test/testfiles/jpp-test_config_0202.jsonp

With latest JsonPreprocessor code changes the error message changed to:

`'Could not set variable 'list_variant_numbersVAR-${variant_str([0]}_MS-${milestone_'2')}_param' with value 'value'! Reason: invalid syntax (<string>, line 1)'`

I wonder about this extension: `Reason: invalid syntax (<string>, line 1)`

Why this? This is not helpful. What is wrong with the syntax? Out of this error message no user will be able to get useful guidance about how to solve the problem.

This test case has no syntax issue. And there is nothing wrong with variables.

The error is that the JsonPreprocessor still is not able to resolve dollar operator expressions within strings correctly.

The resolved variable name is expected to be: `VAR-VARIANT-1_MS-milestone_2_param`.

**Comment by namsonx** [comment]
Hello Holger,

we discussed about the requirement which handling nested parameter in a key name, but it is still not confirmed yet.
The scenario of test case jpp-test_config_0202.jsonp is not implemented yet.

For the reason in error message log, I will remove for a case it has no meaning.

Thank you,
Son

**Comment by namsonx** [comment]
Hello Holger,

During working on the requirement that you defined in test case JPP_0202, I observed the mistake in expected result:
```
[DICT] (22/1) > {variant_1} [STR]  :  'VARIANT-1'
[DICT] (22/2) > {variant_2} [STR]  :  'VARIANT-2'
[DICT] (22/3) > {list_variant_numbers} [LIST] (2/1) > [INT]  :  1
[DICT] (22/3) > {list_variant_numbers} [LIST] (2/2) > [INT]  :  2
[DICT] (22/4) > {list_variant_index} [INT]  :  0
[DICT] (22/5) > {variant_number} [STR]  :  '1'
[DICT] (22/6) > {dict_variants} [DICT] (2/1) > {variant_1} [STR]  :  'VARIANT-1'
[DICT] (22/6) > {dict_variants} [DICT] (2/2) > {variant_2} [STR]  :  'VARIANT-2'
[DICT] (22/7) > {variant_key_1} [STR]  :  'variant_1'
[DICT] (22/8) > {INNERMOST_VARIANT_VALUE} [STR]  :  'RIA'
[DICT] (22/9) > {ARIAN} [STR]  :  'ARIAN'
[DICT] (22/10) > {VARIANT} [STR]  :  'VARIANT'
[DICT] (22/11) > {milestone_1} [STR]  :  'MILESTONE-1'
[DICT] (22/12) > {milestone_2} [STR]  :  'MILESTONE-2'
[DICT] (22/13) > {milestone_3} [STR]  :  'MILESTONE-3'
[DICT] (22/14) > {list_milestone_numbers} [LIST] (3/1) > [INT]  :  1
[DICT] (22/14) > {list_milestone_numbers} [LIST] (3/2) > [INT]  :  2
[DICT] (22/14) > {list_milestone_numbers} [LIST] (3/3) > [INT]  :  3
[DICT] (22/15) > {list_milestone_index} [INT]  :  1
[DICT] (22/16) > {milestone_number} [STR]  :  '2'
[DICT] (22/17) > {dict_milestones} [DICT] (3/1) > {milestone_1} [STR]  :  'MILESTONE-1'
[DICT] (22/17) > {dict_milestones} [DICT] (3/2) > {milestone_2} [STR]  :  'MILESTONE-2'
[DICT] (22/17) > {dict_milestones} [DICT] (3/3) > {milestone_3} [STR]  :  'MILESTONE-3'
[DICT] (22/18) > {milestone_key_2} [STR]  :  'milestone_2'
[DICT] (22/19) > {innermost_milestone_name} [STR]  :  'est'
[DICT] (22/20) > {ilesto} [STR]  :  'ilesto'
[DICT] (22/21) > {milestone} [STR]  :  'milestone'
[DICT] (22/22) > {VAR-VARIANT-1_MS-milestone_2_param} [STR]  :  'value'
```
The string 'milestone' in last line [DICT] (22/22) > {VAR-VARIANT-1_MS-milestone_2_param} [STR]  :  'value' must be uppercase instead of lowercase.
The expected result must be  ```[DICT] (22/22) > {VAR-VARIANT-1_MS-MILESTONE-2_param} [STR]  :  'value'```
It took me a lot of time to recognize this mistake :)

Thank you,
Son

**Comment by HolQue** [comment]
Hi Son,

in case of failing test cases you should take a look at the error messages in the self test log file:

```
Found deviating return values
(1) '[DICT] (22/22) > {VAR-VARIANT-1_MS-milestone-2_param} [STR]  :  'value''   > (expected)
(2) '[DICT] (22/22) > {VAR-VARIANT-1_MS-MILESTONE-2_param} [STR]  :  'value''   > (returned)
```

**Comment by HolQue** [comment]
Hi Son,

test case JPP_0203 has the same issue. But after fixing the capital letters I got the following error message:

`'The variable '${list_milestone_numbers}[1]' is not available!'`

Why this?

All values are the same than in JPP_0202, but in dotdict notation. Therefore I expect the same result as in JPP_0303.

Is something wrong with the dotdict notation in this test case?

**Comment by namsonx** [comment]
Hello Holger,

You are right! I was not aware the dotdict notation in this case. 
I'm working on it, and will inform you once completed.

Thank you,
Son

**Comment by namsonx** [comment]
Hello Holger,

I implemented the nested handling for the dotdict notation in test case ID JPP_0203 in the commit [c4e2d9ae](https://github.com/test-fullautomation/python-jsonpreprocessor/commit/c4e2d9aeb3ac6c422684153bb4083f5beaafa57e) on stabi branch ([namsonx/task/stabi_branch](https://github.com/test-fullautomation/python-jsonpreprocessor/tree/namsonx/task/stabi_branch))

Thank you,
Son

**Comment by HolQue** [comment]
Retest successful

**Comment by test-fullautomation** [comment]
solved and already merged #147
